File: src/main/java/eu/codetopic/utils/ui/activity/navigation/NavigationFragment.java
/*
 * Copyright 2017 Jiří Kuchyňka (Anty)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package eu.codetopic.utils.ui.activity.navigation;

import eu.codetopic.utils.ui.view.holder.loading.LoadingFragment;
import eu.codetopic.utils.ui.view.holder.loading.LoadingVH;

public abstract class NavigationFragment extends LoadingFragment {

    public NavigationFragment() {
    }

    public NavigationFragment(Class<? extends LoadingVH> loadingViewHolderClass) {
        super(loadingViewHolderClass);
    }

    protected NavigationActivity getNavigationActivity() {
        return (NavigationActivity) getActivity();
    }

    protected void switchFragment(Class<? extends NavigationFragment> clazz) {
        getNavigationActivity().replaceFragment(clazz);
    }

    protected void removeSelfFragment() {
        getNavigationActivity().removeCurrentFragment();
    }

}

File: src/main/java/eu/codetopic/utils/ui/activity/navigation/NavigationFragment.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.ui.activity.navigation

import android.support.v4.app.Fragment
import android.util.Log
import eu.codetopic.utils.ui.view.holder.loading.LoadingFragment
import eu.codetopic.utils.ui.view.holder.loading.LoadingVH

abstract class NavigationFragment : LoadingFragment {

    companion object {

        private const val LOG_TAG = "NavigationFragment"
    }

    protected open val navigationActivity: NavigationActivity?
        get() = activity.takeIf { it is NavigationActivity? } as NavigationActivity?

    constructor()

    constructor(loadingViewHolderClass: Class<out LoadingVH>) : super(loadingViewHolderClass)

    protected open fun <T : Fragment> switchFragment(clazz: Class<out T>): T? {
        navigationActivity?.apply { return replaceFragment(clazz) } ?:
                Log.e(LOG_TAG, "switchFragment(clazz=$clazz) -> " +
                        "Failed to switch fragment: No NavigationActivity available.")
        return null
    }

    protected open fun removeSelf() =
            navigationActivity?.apply { removeCurrentFragment() } ?:
                    Log.e(LOG_TAG, "removeSelf() -> " +
                            "Failed to remove fragment: No NavigationActivity available.")

}
